### DISCUSSION TITLE
Support for Teensy 4 and SPI transaction

### DISCUSSION THREAD

**Discussion by k-ljung** [RudolphRiedel/FT800-FT813#91] [Ideas]
Hi @RudolphRiedel 

I have created a branch where I added support for Teensy 4 and also implemented support different SPI speed when sharing the SPI bus with other components. At the moment I only implemented the SPI changes for the Teensy target.

https://github.com/RudolphRiedel/FT800-FT813/compare/5.x...k-ljung:FT800-FT813:5.x

Missed the Teensy 4 configuration in PlatformIO
https://github.com/RudolphRiedel/FT800-FT813/commit/e1af02d6925879615d2b8ad82d1ea6f3bb9f2ee4

Removed wrong SPI calls
https://github.com/k-ljung/FT800-FT813/commit/f0145f7d45509788554cdb044191b4268a3bb923

**Comment by RudolphRiedel** [comment]
Ok, to make it compile for the Teensy 4.0 as well just three additional " || (ARDUINO_TEENSY40)" are needed in three files, EVE_target.h, EVE_target_Aduino_Teensy4.h and EVE_cpp_target.cpp.
I know you figured that out, but I wanted to document this for the benefit of anyone else who reads this in the future. :-)

So I just did this locally and also added a block to platformio.ini and it compiles.
Ok, I will push that change.

The change with SPI.beginTransaction() / SPI.endTransaction() is more complicated in what the implications are.
This buries the settings deep in the library files and the SPI clock does not only depend on the actual hardware,
using a fixed 30MHz violates the requirement that the clock during initialization must not be higher than 11MHz.
I am not even sure if I could use 30MHz after init with my setup and the Teensy 4.x boards.
And using SPI.beginTransaction() / SPI.endTransaction() for every single SPI transfer is a lot of overhead that slows everything down.

So a better idea would be to put that in the user space and use
SPI.beginTransaction(SPISettings(8000000, MSBFIRST, SPI_MODE0));
TFT_init();
SPI.endTransaction();
SPI.beginTransaction(SPISettings(1000000, MSBFIRST, SPI_MODE1));
somethingelse();
SPI.endTransaction();

But then there is the issue that a number of targets, including Teensy 4x is using DMA with the _burst() functions.
So this:
SPI.beginTransaction(SPISettings(1600000, MSBFIRST, SPI_MODE0));
TFT_display();
SPI.endTransaction();

Either is blocking untill the DMA is done which really defeats the purpose of using DMA, or it will not work at all.

So code using additional SPI nodes should be using a state machine that is using EVE_busy() to determine if a transfer is still ongoing and only switch states to use a different SPI node if EVE_busy() returns E_OK.
There also is the global variable EVE_dma_busy that is set to zero when the DMA transfer is done.
